File: compiler/bench/issorted.fs
\ expected: 5000
\ Check if array is sorted benchmark - 10K iterations

1000 constant SIZE
10000 constant ITERS
create arr SIZE cells allot

: arr@ ( i -- n ) cells arr + @ ;
: arr! ( n i -- ) cells arr + ! ;

: is-sorted? ( -- flag )
  1                       \ assume sorted (descending)
  SIZE 1 do
    i 1- arr@ i arr@ < if
      drop 0 unloop exit
    then
  loop ;

: init-sorted ( -- )
  SIZE 0 do SIZE i - 1- i arr! loop ;

: init-unsorted ( -- )
  SIZE 0 do i 7 * SIZE mod i arr! loop ;

: bench ( -- count )
  0
  ITERS 0 do
    i 2 mod if init-unsorted else init-sorted then
    is-sorted? if 1+ then
  loop ;

: main
  bench . cr ;

File: compiler/bench/issorted.c
// expected: 5000
// Check if array is sorted benchmark - 10K iterations

#include <stdio.h>

#define SIZE 1000
#define ITERS 10000

int arr[SIZE];

int is_sorted(void) {
    for (int i = 1; i < SIZE; i++) {
        if (arr[i - 1] < arr[i]) return 0;
    }
    return 1;
}

void init_sorted(void) {
    for (int i = 0; i < SIZE; i++) {
        arr[i] = SIZE - i - 1;
    }
}

void init_unsorted(void) {
    for (int i = 0; i < SIZE; i++) {
        arr[i] = (i * 7) % SIZE;
    }
}

int bench(void) {
    int count = 0;
    for (int i = 0; i < ITERS; i++) {
        if (i % 2) init_unsorted();
        else init_sorted();
        if (is_sorted()) count++;
    }
    return count;
}

int main(void) {
    printf("%d\n", bench());
    return 0;
}
